Add new NativeCurlClient (#262)
kbsali/php-redmine-api: docs/migrate-to-nativecurlclient.md

> # Migrate from `Redmine\Client` to `Redmine\Client\NativeCurlClient`

Reviewer: ```suggestion
# Migrating from `Redmine\Client` to `Redmine\Client\NativeCurlClient`
```

> Since `php-redmine-api` v1.8.0 there is a new native cURL client
> `Redmine\Client\NativeCurlClient`. This guide will help you to migrate your code
> to this client.

Reviewer: ```suggestion
Since `v1.8.0`, there is a new native cURL client `Redmine\Client\NativeCurlClient`. This guide will help migrate your code to this client.
```

> ## 1. Use new client methods

Reviewer: ```suggestion
## 1. New client's methods
```

> With the new interface `Redmine\Client\Client` there are now standarized methods
> for all clients. The new `Redmine\Client\NativeCurlClient` and the current `Redmine\Client` implementing this interface.

Reviewer: ```suggestion
With the new interface `Redmine\Client\Client` there are now standardized methods for all clients. The new `Redmine\Client\NativeCurlClient` and the current `Redmine\Client` both implement this interface.
```

> ### api() to getApi()

Reviewer: ```suggestion
### `api()` -> `getApi()`
```

> Search in your code for the usage of `$client->api('issue')` and the magic
> getter like `$client->issue`. Then replace this calls with
> `$client->getApi('issue')`.

Reviewer: ```suggestion
We got rid of `api()` method as well as the magic getters. So now to query the `issue` API for instance, use `$client->getApi('issue')` instead. See the following example :
```

```diff
-$issue = $client->issue->show($issueId);
+$issue = $client->getApi('issue')->show($issueId);

-$client->api('issue')->create($data);
+$client->getApi('issue')->create($data);
```

> ### getResponseCode() to getLastResponseStatusCode()

Reviewer: ```suggestion
### `getResponseCode()` -> `getLastResponseStatusCode()`
```

> Replace every call for `$client->getResponseCode()` with
> `$client->getLastResponseStatusCode()`.

Reviewer: ```suggestion
Replace every calls for `$client->getResponseCode()` with `$client->getLastResponseStatusCode()`.
```

```diff
-if ($client->getResponseCode() === 500)
+if ($client->getLastResponseStatusCode() === 500)
{
    throw new \Exception('Redmine call failed');
}
```

> ### get() to requestGet()

Reviewer: ```suggestion
### `get()` -> `requestGet()`
```
But is there a reason for doing this? I liked the simple approach of `$client->get()`,  `$client->post()`, `$client->put()` or `$client->delete()`

Author: I liked the shorter methods too. But the reason is backward compatibility. `get()` doesn't have typed parameters and returns () `string`, `array` or `false`. For forward compatibility we need new methods which I named `request{Get|Post|...}` in lack of better ideas. `requestGet()` has typed parameters and only returns bool. We could rename them back in v2 but this would destroy our work on the forward compatibility.

> If you are using `$client->get()`, `$client->post()`, `$client->put()` or
> `$client->delete()` directly you will have to change your code. This methods
> parse a possible JSON or XML response but in future the parsing of the raw
> response body will be up to you.

Reviewer: ```suggestion
If you are using `$client->get()`, `$client->post()`, `$client->put()` or `$client->delete()` directly you will have to change your code. Currently, these methods parse the response (JSON or XML), but in future the parsing of the raw response body will be up to you.
```
(could we make all the paragraph a long line? Or is there a reason for having line breaks at the 80th chars?)

Author: I've noted that that shorter lines are much more useful in git diffs. But we could also got with long lines.

To help you with the parsing of the raw response the client interface introduces
two new methods: `getLastResponseContentType()` and `getLastResponseBody()`.

This example shows how you can parse the response body of a GET request.

```diff
-// We dont know if we will get json, xml or a string
-$dataAsJsonOrXmlOrString = $this->client->get($path);
+$this->client->requestGet($path);
+// $body contains the raw http body of the response
+$body = $this->client->getLastResponseBody();
+
+// if response is XML, create a SimpleXMLElement object

> +if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/xml')) {

Reviewer: ok for me, but maybe we should add another method that does exactly this, what do you think? It makes it more complex to use the lib that way...

Author: This code is copied from the `AbstractApi::get()` method. Parsing would only needed if someone don't use the `AbstractApi` but calls `$client->get()` directly. Personally I don't think that someone has done this but I think it's worth mentioning.

+    $dataAsXML = new \SimpleXMLElement($body);
+} else if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/json')) {
+    try {
+        $dataAsJson = json_decode($body, true, 512, \JSON_THROW_ON_ERROR);
+    } catch (\JsonException $e) {
+        throw new \Exception('Error decoding body as JSON: '.$e->getMessage());
+    }
+} else {
+    $dataAsString = $body;
+}
```

> ### post() to requestPost()

Reviewer: ```suggestion
### `post()` -> `requestPost()`
```

Reviewer: same for the following headers

+    $dataAsXML = new \SimpleXMLElement($body);
+} else {
+    $dataAsString = $body;
+}
```

### put() to requestPut()

This example shows how you can parse the response body of a PUT request.

```diff
-// We dont know if we will get xml or a string
-$dataAsXmlOrString = $this->client->put($path, $data);
+$this->client->requestPut($path, $data);
+// $body contains the raw http body of the response
+$body = $this->client->getLastResponseBody();
+
+// if response is XML, create a SimpleXMLElement object
+if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/xml')) {
+    $dataAsXML = new \SimpleXMLElement($body);
+} else {
+    $dataAsString = $body;
+}
```

### delete() to requestDelete()

This example shows how you can parse the response body of a DELETE request.

```diff
-$dataAsString = $this->client->delete($path);
+$this->client->requestDelte($path);
+$dataAsString = $this->client->getLastResponseBody();
```

### setImpersonateUser() to startImpersonateUser()

If you are using the [Redmine user impersonation](https://www.redmine.org/projects/redmine/wiki/Rest_api#User-Impersonation)

> you have to change your code.

Reviewer: ```suggestion
you have to change your code as follow :
```

```diff
// impersonate the user `robin`
-$client->setImpersonateUser('robin');

> +$client->startImpersonateUser('robin');

Reviewer: ```suggestion
+$client->startUserImpersonation('robin');
```

Author: Personally :+1: but since `startImpersonateUser()` is now part of the released client interface we can not change this without breaking change.

$userData = $client->getApi('user')->getCurrentUser();

// Now stop impersonation
-$client->setImpersonateUser(null);

> +$client->stopImpersonateUser();

Reviewer: ```suggestion
+$client->stopUserImpersonation();
```

Author: Same here: `stopImpersonateUser()` is already part of the released client interface.

```

### setCheckSslCertificate() to setCurlOption()

```diff
-$client->setCheckSslCertificate(true);
+$client->setCurlOption(CURLOPT_SSL_VERIFYPEER, true);
```

### setCheckSslHost() to setCurlOption()

```diff
-$client->setCheckSslHost(true);

> +$client->setCurlOption(CURLOPT_SSL_VERIFYHOST, 2);

Reviewer: again, this is a bit to verbose to me. It means the developer is supposed to know about the `CURLOPT_SSL_VERIFYHOST -> 2` detail. Abstracting these kind of details feels like good for the DX (Developer Experience)

Author: I'm not sure if having a special method for specific cURL options leads to better DX. I one searches for specific cURL features they will end up on SO or curl.se with instructions to set a specific cURL options. How will one know that there is a specific method instead of setting the cURL option directly with `setCurlOption()`?
Abstracting cURL features will made us care for them. Also where would be the line for which cURL features we should add and which not?
Since this is a Redmine library I would only care about abstract Redmine features (like user impersonation, auth, etc) but leaving the http connection part out of scope of the project.

Authorization for Redmine uses always the http headers `X-Redmine-API-Key` or
`Authorization`. So the setting `setUseHttpAuth()` is not longer needed.

```diff
-$client->setUseHttpAuth(false);
```

If you want to override the `X-Redmine-API-Key` or `Authorization` header you
can set them via `setCurlOption()` method.

```diff
-$client->setUseHttpAuth(false);
+$client->setCurlOption(CURLOPT_HTTPHEADER, ['Authorization: Basic username_password_base64_encoded']);
// or
+$client->setCurlOption(CURLOPT_HTTPHEADER, ['X-Redmine-API-Key: secret_access_key']);
```

### setPort() to setCurlOption()

```diff
-$client->setPort(8080);
+$client->setCurlOption(CURLOPT_PORT, 8080);
```

### getResponseCode() to getLastResponseStatusCode()

```diff
-$statusCode = $client->getResponseCode();
+$statuscode = $client->getLastResponseStatusCode();
```

### setImpersonateUser() to startImpersonateUser()

```diff
// impersonate the user `robin`
-$client->setImpersonateUser('robin');
+$client->startImpersonateUser('robin');

$userData = $client->getApi('user')->getCurrentUser();

// Now stop impersonation
-$client->setImpersonateUser(null);
+$client->stopImpersonateUser();

> ```

Reviewer: already mentioned above

### setCustomHost() to setCurlOption()

```diff
-$client->setCustomHost('http://custom.example.com');
+$client->setCurlOption(CURLOPT_HTTPHEADER, ['Host: http://custom.example.com']);
```

> ## 2. Stop using deprecated methods

Reviewer: ```suggestion
## 2. Methods deprecation
```

The following methods are deprecated and were set public or protected only for
testing. They are not available in `NativeCurlClient`. If you are using them,
please remove them.

- `getUrl()`
- `decode()`
- `getCheckSslCertificate()`
- `getCheckSslHost()`
- `getSslVersion()`
- `getUseHttpAuth()`
- `getPort()`
- `getImpersonateUser()`
- `getCustomHost()`
- `getCurlOptions()`
- `prepareRequest()`
- `processCurlResponse()`
- `runRequest()`

## 3. Switch to `NativeCurlClient`

> As the old client the `Redmine\Client\NativeCurlClient` requires:

Reviewer: ```suggestion
Just like with the old `Redmine\Client`, the `Redmine\Client\NativeCurlClient` requires :
```

Author: Is having a space before `:` common practice? Personally I havn't seen this before.

- a URL to your Redmine instance
- an Apikey or username
- and optional a password if you want to use username/password (not recommended).

> So after you made all changes in the previous sections you should be able to

Reviewer: ```suggestion
So after making all the changes suggested in the previous sections you should be able to
```

test your code without errors and now simply switch the client.

```diff
// Instantiate with ApiKey
-$client = new \Redmine\Client(
+$client = new \Redmine\Client\Prs18Client(
    'https://redmine.example.com',
    '1234567890abcdfgh'
);
```

Now you should be ready. Please make sure that you are only using client public

> methods that are defined in `Redmine\Client\NativeCurlClient` because all other

Reviewer: ```suggestion
methods that are defined in `Redmine\Client\NativeCurlClient` because all the other
```

methods will be removed or set to private in the next major release. Otherwise

> you will have to change your code in future again.

Reviewer: ```suggestion
you will have to change your code in the future again.
```
